include/boost/geometry/algorithms/detail/overlay/get_turn_info_for_endpoint.hpp

```diff
@@ -279,37 +317,25 @@ struct get_turn_info_for_endpoint
         return result_ignore_ip0 || result_ignore_ip1;
     }
 
-    template <typename Point1,
-              typename Point2,
+    template <typename UniqueSubRange1,
+              typename UniqueSubRange2,
               typename TurnInfo,
               typename IntersectionInfo,
               typename OutputIterator>
     static inline
-    bool analyse_segment_and_assign_ip(Point1 const& pi, Point1 const& pj, Point1 const& pk,
-                                       Point2 const& qi, Point2 const& qj, Point2 const& qk,
-                                       bool is_p_first, bool is_p_last,
-                                       bool is_q_first, bool is_q_last,
+    bool analyse_segment_and_assign_ip(UniqueSubRange1 const& range_p,
+                                       UniqueSubRange2 const& range_q,
                                        linear_intersections::ip_info const& ip_info,
                                        TurnInfo const& tp_model,
                                        IntersectionInfo const& inters,
                                        unsigned int ip_index,
                                        OutputIterator out)
     {
-#ifdef BOOST_GEOMETRY_DEBUG_GET_TURNS_LINEAR_LINEAR
-        // may this give false positives for INTs?
-        typename IntersectionResult::point_type const&
-            inters_pt = result.template get<0>().intersections[ip_index];
-        BOOST_GEOMETRY_ASSERT(ip_info.is_pi == equals::equals_point_point(pi, inters_pt));
-        BOOST_GEOMETRY_ASSERT(ip_info.is_qi == equals::equals_point_point(qi, inters_pt));
-        BOOST_GEOMETRY_ASSERT(ip_info.is_pj == equals::equals_point_point(pj, inters_pt));
-        BOOST_GEOMETRY_ASSERT(ip_info.is_qj == equals::equals_point_point(qj, inters_pt));
-#endif
-
         // TODO - calculate first/last only if needed
-        bool is_p_first_ip = is_p_first && ip_info.is_pi;
-        bool is_p_last_ip = is_p_last && ip_info.is_pj;
-        bool is_q_first_ip = is_q_first && ip_info.is_qi;
-        bool is_q_last_ip = is_q_last && ip_info.is_qj;
+        bool is_p_first_ip = range_p.is_first() && ip_info.is_pi;
+        bool is_p_last_ip = range_p.size() == 2u && ip_info.is_pj;
```

Comment:
This condition looks non-intuitively. Is it based on an assumption that for last segment the point `k` is not retrieved? This way here, the caller has to know how the `range_p` and `range_q` below are filled possibly in some other place of the code. Would it be possible to have and explicit function `is_last()` (like `is_first()`) so here it'd be clear what's happening and all details are hidden in the `range_p`?